File: 662B.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef vector<int> vi;

string solve(unordered_map<int, int> count) {
    int a, b, c;
    a = b = c = 0;
    for (auto elem : count) {
        int cur = elem.second;
        
        if (cur <= 1) continue;

        if (a == 0 && cur >= 4) {
            a = 4;
            cur -= 4;
        }
        if (b == 0 && cur >= 2) {
            b = 2;
            cur -= 2;
        }
        if (c == 0 && cur >= 2) {
            c = 2;
            cur -= 2;
        }

        if (a == 4 && b == 2 && c == 2) {
            return "YES\n";
        }
    }

    return "NO\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

	int n; cin >> n;
    int plank;
    string symbol;
    unordered_map<int, int> count;
    for (int i = 0; i < n; ++i) {
        cin >> plank;
        count[plank]++;
    }
    int m; cin >> m;
    string res = "";
    vi freq(4, 0); 8, 6, 2, 2
    for (int i = 0; i < m; ++i) {
        cin >> symbol >> plank;
        if (symbol == "+")
            count[plank]++;
        else if (symbol == "-")
            count[plank]--;
        if (count[plank] >= 8)
        res += solve(freq);
    }
    cout << res;
}

File: 662B.py
import collections

n = int(input())
s = list(map(int, input().split()))
count = collections.Counter(s)

def solve(count):
    a, b, c = None, None, None
    for key in sorted(count.keys(), reverse=True):
        val = count[key]
        if val >= 8:
            return "YES"
        elif val >= 6:
            if not a:
                a, b = key, key
            else:
                return "YES"
        elif val >= 4:
            if not a:
                a = key
            else:
                return "YES"
        elif val >= 2:
            if not b:
                b = key
            else:
                return "YES"
    
    return 'NO'

m = int(input())
res = []
for _ in range(m):
    tmp = input().split()
    if tmp[0] == "+":
        count[int(tmp[1])] += 1
    elif tmp[0] == "-":
        count[int(tmp[1])] -= 1

    res.append(solve(count))

for s in res:
    print(s)